### ISSUE TITLE
[cloud_firestore] crashes when connection turns off

### ISSUE THREAD

**Issue by kroikie** [FirebaseExtended/flutterfire#763]
Hi. I have faced an unexpected behavior of Firestore persistence. The problem is that it throws an error when you're running your app and turn off WiFi. After this all database related commands will suspend your app. However, when you restart the app without internet, it seems that all changes have been done.

## Steps to Reproduce

1. Do anything like this:
```dart
Firestore.instance.collection(collection).document().setData(data);
```
or this:
```dart
FutureBuilder<FirebaseUser>(
  future: Auth.currentUser(),
  builder: (context, snapshot) {
    if (snapshot.connectionState == ConnectionState.done) {
      return StreamBuilder<QuerySnapshot>(
        stream: Firestore.instance
                         .collection('collection')
                         .where('owner', isEqualTo: snapshot.data.uid)
                         .snapshots(),
        builder: (context, snapshot) {
          if (!snapshot.hasData) {
            return Center(child: CircularProgressIndicator());
          }
          return ListView.builder(
            itemCount: snapshot.data.documents.length,
            itemBuilder: (context, index) {
              var document = snapshot.data.documents[index];
              var documentId = document.documentID;
                      // Display your content somehow
```
2. Run your app and see some data fetched from server
3. Disable the Internet

## Logs

```
V/NativeCrypto(11784): Read error: ssl=0x9e50e908: I/O error during system call, Connection timed out
W/zzaeo   (11784): [{0}] Failed to resolve name. status={1}
V/NativeCrypto(11784): Write error: ssl=0x9e50e908: I/O error during system call, Broken pipe
V/NativeCrypto(11784): Write error: ssl=0x9e50e908: I/O error during system call, Broken pipe
W/zzajf   (11784): Failed closing connection
W/zzajf   (11784): javax.net.ssl.SSLException: Write error: ssl=0x9e50e908: I/O error during system call, Broken pipe
W/zzajf   (11784): 	at com.google.android.gms.org.conscrypt.NativeCrypto.SSL_write(Native Method)
W/zzajf   (11784): 	at com.google.android.gms.org.conscrypt.NativeSsl.write(:com.google.android.gms@12685022@12.6.85 (040308-197041431):2)
W/zzajf   (11784): 	at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket$SSLOutputStream.write(:com.google.android.gms@12685022@12.6.85 (040308-197041431):6)
W/zzajf   (11784): 	at com.google.android.gms.internal.firebase-firestore.zzapc.write(Unknown Source)
W/zzajf   (11784): 	at com.google.android.gms.internal.firebase-firestore.zzaot.write(Unknown Source)
W/zzajf   (11784): 	at com.google.android.gms.internal.firebase-firestore.zzapf.close(Unknown Source)
W/zzajf   (11784): 	at com.google.android.gms.internal.firebase-firestore.zzaku.close(Unknown Source)
W/zzajf   (11784): 	at com.google.android.gms.internal.firebase-firestore.zzaim.run(Unknown Source)
W/zzajf   (11784): 	at com.google.android.gms.internal.firebase-firestore.zzahi.run(Unknown Source)
W/zzajf   (11784): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
W/zzajf   (11784): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
W/zzajf   (11784): 	at java.lang.Thread.run(Thread.java:818)
W/zzaeo   (11784): [{0}] Failed to resolve name. status={1}
```

```
info • 'debugEnableFadingRoutes' is deprecated and shouldn't be used • lib/main.dart:7:21
info • The method '_signedOut' isn't used • lib/root.dart:33:8

2 issues found. (ran in 10.9s)
```

```
[✓] Flutter (Channel beta, v0.5.1, on Linux, locale en_US.UTF-8)
    • Flutter version 0.5.1 at /home/alex/flutter
    • Framework revision c7ea3ca377 (5 weeks ago), 2018-05-29 21:07:33 +0200
    • Engine revision 1ed25ca7b7
    • Dart version 2.0.0-dev.58.0.flutter-f981f09760

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.0)
    • Android SDK at /home/alex/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.0
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 3.1)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 26.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (1 available)
    • Nexus 5 • 0226d62443a398a7 • android-arm • Android 6.0.1 (API 23)

• No issues found!
```

**Comment by kroikie** [comment]
@alexbagirov

The issue at https://github.com/flutter/flutter/issues/19122 has been closed and moved here. Future collaboration on this issue will be done here.

**Comment by iapicca** [comment]
Hi @alexbagirov
If you are still experiencing this issue 
with the latest version of Flutter and FlutterFire plugin
can you please provide your updated  `flutter doctor -v`
and your `flutter run --verbose`/`flutter build --verbose` ?
Also, to better address the issue, would be helpful
if you could post a self contained app on github
or the steps to reproduce it.
Thank you

**Comment by PrismaSoftGC** [comment]
Good afternoon, I have the following situation in firebase_firestore. I need to save data when I am offline, but I do 2 queries because before saving, I seek the correct data in another collection. So far it's working normally, but if I hit the back button until I close the app and then turn on the internet to sync the data, nothing just happens. The odd thing is that if the application remains open or minimized, it works normal. Can anyone help me?

`
Firestore.instance
.collection ('Region')
.where ('idRegion', isEqualTo: idRegion)
.snapshots ()
.listen ((date) {
data.documents.forEach ((doc) {

Firestore.instance
.collection ('NotebookDefensive Field')
.document (idCadernoCampo)
.updateData ({'region': doc ['RegionName']})
.then ((_) {});

});
});
`

**Comment by helenaford** [comment]
Hi @PrismaSoftGC , if you are still experiencing this issue
with the latest version of Flutter and FlutterFire plugin
can you please provide your updated flutter doctor -v
and your flutter run --verbose/flutter build --verbose ?
Also, to better address the issue, would be helpful
if you could post a self contained app on github
or the steps to reproduce it.
Thank you

**Comment by Salakar** [comment]
Hey, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) (native exceptions are now properly caught without crashing and propagated to Dart) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
